Refactor packages and try to implement parallel calculations

bend-is opened the pull request:
An attempt to implement parallel computation for task 3

bend-is committed:
Refactor packages and try to implement parallel calculations

korjavin commented on task3/pkg/textprocessor/textprocessor.go:
```diff
@@ -27,43 +29,68 @@ func New(reader io.Reader, wordLength int) *TextProcessor {
 		if line == "" {
 			continue
 		}
+		wg.Add(1)
```
Отлично, правильное использование wg

korjavin commented on task3/pkg/textprocessor/textprocessor.go:
```diff
@@ -27,43 +29,68 @@ func New(reader io.Reader, wordLength int) *TextProcessor {
-				if utf8.RuneCountInString(word) <= tp.wordLength {
-					continue
+					// Magic 100 to keep order.
+					sMap.IncrementCount(word, lineNumber*100+wordPosition)
```
100 может не сработать, и в любом случае надо убратьв  константы

bend-is replied:
Ну 100 не сработает если в строке будет больше чем 100 слов. Вынес в константу и увеличил до 1000. Но опять же могут быть строки в которых больше 1000 слов. Не могу придумать как правильней поддерживать порядок появления слов.

korjavin replied:
Можно иметь два числа. Тогда точно не пересекутся

bend-is replied:
Спасибо, так и сделал!

bend-is committed:
Use channel to sync adding a stop words.
Added goroutine run count limit for textprocessor.
Remove all mutex operation from sortedmap.

korjavin commented on task3/pkg/textprocessor/textprocessor.go:
```diff
@@ -11,28 +11,36 @@ import (
+	// Need to keep order. Thinking that can't be more then 1000 words in line.
+	orderMultiplier = 1000
+	// Number of max goroutines.
+	maxGoroutines = 12
```
Лучше в параметр

bend-is replied:
Вынес

korjavin commented on task3/pkg/textprocessor/textprocessor.go:
```diff
@@ -11,28 +11,36 @@ import (
 	scanner := bufio.NewScanner(tp.source)
 	reg := regexp.MustCompile(`[^a-zA-Z]`)
 
+	wordCh, stopCh := tp.listenStopWords(sMap)
```
А не хочется метод listen иметь в map а не в tp ? просто рассуждаю.

bend-is replied:
Избавился вобще от него

korjavin commented on task3/pkg/textprocessor/textprocessor.go:
```diff
@@ -42,8 +50,11 @@ func (tp *TextProcessor) CountWords() *sortedmap.SortedMap {
+		guardCh <- struct{}{}
 		go func(lineNumber int) {
 			defer wg.Done()
+			defer func() { <-guardCh }()
```
мы можем объеденить defer

bend-is replied:
Сделал

korjavin commented on task3/pkg/sortedmap/sortedmap.go:
```diff
@@ -1,46 +1,46 @@
 
-func (s *SortedMap) Add(item string) {
+func (s *SortedMap) AddStopWord(word string) {
+	s.stopWords[word] = struct{}{}
```
SortedMap не threadSafe в публичном интерфейсе. Опасно.

bend-is replied:
Обезопасил

korjavin commented on task3/pkg/textprocessor/textprocessor.go:
```diff
@@ -5,65 +5,105 @@ import (
+			select {
+			case word := <-wordCh:
+				sMap.AddStopWord(word)
+			case <-stopCh:
```
Без этого канала можно обойтись закрывая канал wordCh

bend-is replied:
Вернулся к мьютексам

bend-is committed:
Refactor after code review
